test/integration/targets/changed_when/tasks/main.yml

```diff
@@ -59,3 +59,29 @@
   assert:
     that:
       - groupby is not changed
+
+- add_host:
+    name: 'host_{{item}}'
+  loop:
+    - 1
+    - 2
+  changed_when: item == 2
```

Comment:
I've checked out your branch and this is working. However, I still want to note that there is a behavior change where `add_host` used to give "changed" and now gives "ok". I understand that there may be some philosophical reasons for this, like how the _target_ host isn't actually changed. That's not as important as fixing the error, and is probably out of scope for this fix.

Reply:
> there is a behavior change where add_host used to give "changed" and now gives "ok"

Do you know when did that behavior change happen? Was the `changed` value correct before or after? There is a PR (merged) that fixed the `changed` value: https://github.com/ansible/ansible/pull/32057.

Reply:
In 2.9, bc955b9d6011c432a941ce7cfee2567fa23f238b, it gave "changed". I will do a git bisect.

That leaves me with efe103cdff39326f27e12616d3cee76e7f30226c as the commit that the behavior change was introduced, PR https://github.com/ansible/ansible/pull/69897

I don't think the behavior was correct _before_ that PR either. It went from _always_ giving changed to _always_ giving ok. Neither of these seem right, consider:

```yaml
    - add_host:
        name: 'host_{{item}}'
      loop:
        - 1
        - 2
        - 2
```

I would expect changed-changed-ok. The reason is that the first 2 entries in the loop are unique, the 3rd one isn't. When it runs the 3rd entry, it adds a host that already exists.

Reply:
Sorry only now I noticed that it was also part of the original issue. Sounds like that is still specific to loops since the following ("un-looped" version) works correctly:
```yaml
- hosts: localhost
  gather_facts: no
  tasks:
    - add_host:
        name: 'host_1'
    - add_host:
        name: 'host_2'
    - add_host:
        name: 'host_2'
```

Reply:
well that's wild. I can write that up in a new issue if you want to merge this as it is.

Reply:
So the reason why this happens is because callbacks for each item of the loop are issued during the execution of the loop but `add_host` and `group_by` are processed after the loop has ended. So at the time the callbacks are issued the value of `changed` is set to `False` because we don't know the actual value yet.